Lesson_24/js/validator.js

```diff
@@ -0,0 +1,101 @@
+         message: "The field should not contain numbers",
+         errorType: 'NaN',
+      },
+      isEmailValid: {
+         validate: (value) => {
+            const regExpTypeEmail = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
+            return regExpTypeEmail.test(value);
+         },
+         message: "Check the correct writing of mail",
+         errorType: 'mail',
+      },
+      isPasswordSecure: {
+         validate: (value) => {
+            const regExpDifficultyCheck = /^(?=.*[a-z])(?=.*[A-Z])(?=.*[0-9])(?=.*[!@#\$%\^&\*])(?=.{8,})/;
+            return regExpDifficultyCheck.test(value);
+         },
+         message: "Password must has at least 8 characters that include at least 1 lowercase character, 1 uppercase characters, 1 number, and 1 special character in (!@#$%^&*)",
+         errorType: 'passwordSecure',
+      },
+      lengthRange(minLength, maxLength) {
+         return {
+            validate: (value) => value.length >= minLength && value.length <= maxLength,
+            message: `The field must be between ${minLength} and ${maxLength} characters`,
+            errorType: 'lengthRange'
+         }
+      },
+      minLength(length) {
+         return {
+            validate: (value) => value.length >= length,
+            message: `The field must contain at least ${length} characters`,
+            errorType: 'minLength'
+         }
+      },
+      passwordMatching(passName) {
+         return {
+            validate: (value, elements) => {
+               const passValue = elements[passName].value;
+               return value === passValue;
+            },
+            message: "Please enter the password again",
+            errorType: 'passwordMatching'
+         }
+      }
+   },
+
+   validate(form, config) {
+      if (!(form instanceof HTMLFormElement)) {
+         throw new ValidationError('You should provide HTML form');
+      }
+
+      let elements = form.elements;
+      this.errors[form.name] = {};
+
+      for (const [inputName, inputValidators] of Object.entries(config)) {
+         if (!inputValidators.length) {
+            continue;
+         }
+
+         if (!elements[inputName]) {
+            throw new ValidationError(`The "${inputName}" field doesn't exist in the "${form.name}"`);
+         }
+
+         const value = elements[inputName].value.trim();
+         let errors = this.errors[form.name];
+
+         inputValidators.forEach(({ validate, message, errorType }) => {
+            if (!validate(value, elements)) {
+               errors[inputName] = {
+                  ...errors[inputName],
+                  [errorType]: message,
+               }
+            }
+         });
+      }
+
+      return !this._hasError(form.name);
+   },
+
+   getErrors(formName) {
+      return this.errors[formName];
+   },
+
+   _hasError(formName) {
```

Comment:
don't use underscore naming.  It is generally not recommended to use underscore naming of private methods because it is no longer considered a reliable way to indicate that a method is private. Use # for name private methods